flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/SimpleTransformationTranslator.java

```diff
@@ -0,0 +1,126 @@
+	public Collection<Integer> translateForStreaming(final T transformation, final Context context) {
+		checkNotNull(transformation);
+		checkNotNull(context);
+
+		final Collection<Integer> transformedIds =
+				translateForStreamingInternal(transformation, context);
+		configure(transformation, context);
+
+		return transformedIds;
+	}
+
+	/**
+	 * Translates a given {@link Transformation} to its runtime implementation for BATCH-style execution.
+	 *
+	 * @param transformation The transformation to be translated.
+	 * @param context The translation context.
+	 * @return The ids of the "last" {@link StreamNode StreamNodes} in the transformation graph corresponding
+	 * to this transformation. These will be the nodes that a potential following transformation will need to
+	 * connect to.
+	 */
+	protected abstract Collection<Integer> translateForBatchInternal(final T transformation, final Context context);
+
+	/**
+	 * Translates a given {@link Transformation} to its runtime implementation for STREAMING-style execution.
+	 *
+	 * @param transformation The transformation to be translated.
+	 * @param context The translation context.
+	 * @return The ids of the "last" {@link StreamNode StreamNodes} in the transformation graph corresponding
+	 * to this transformation. These will be the nodes that a potential following transformation will need to
+	 * connect to.
+	 */
+	protected abstract Collection<Integer> translateForStreamingInternal(final T transformation, final Context context);
+
+	private void configure(final T transformation, final Context context) {
+		final StreamGraph streamGraph = context.getStreamGraph();
+		final int transformationId = transformation.getId();
+
+		if (transformation.getBufferTimeout() >= 0) {
+			streamGraph.setBufferTimeout(transformationId, transformation.getBufferTimeout());
+		} else {
+			streamGraph.setBufferTimeout(transformationId, context.getDefaultBufferTimeout());
+		}
+
+		if (transformation.getUid() != null) {
+			streamGraph.setTransformationUID(transformationId, transformation.getUid());
+		}
+		if (transformation.getUserProvidedNodeHash() != null) {
+			streamGraph.setTransformationUserHash(transformationId, transformation.getUserProvidedNodeHash());
+		}
+
+		if (!streamGraph.getExecutionConfig().hasAutoGeneratedUIDsEnabled()) {
+			if (transformation instanceof PhysicalTransformation
+					&& transformation.getUserProvidedNodeHash() == null
+					&& transformation.getUid() == null) {
+				throw new IllegalStateException("Auto generated UIDs have been disabled " +
+						"but no UID or hash has been assigned to operator " + transformation.getName());
+			}
+		}
+
+		if (transformation.getMinResources() != null && transformation.getPreferredResources() != null) {
+			streamGraph.setResources(transformationId, transformation.getMinResources(), transformation.getPreferredResources());
+		}
+
+		final StreamNode streamNode = streamGraph.getStreamNode(transformationId);
+		if (streamNode != null
```

Comment:
nit: How about we throw an exception if both are set? Otherwise it might hide some problems, as we mix different memory requirements.

Reply:
You mean throw an exception if one is set and the other is not, right?

Reply:
I mean if both the `translateForBatch/Stream` and the transformation itself define some requirements.

Reply:
I am wondering do we want to fail, or simply give priority to what is set in the `translateForBatch/Stream`? The reason is that I do not know what is done in the Table API with the existing transformations.

Reply:
I can't tell for sure. That's also somewhat where my comment comes from. Currently, it will be hard to track why/which setting takes precedence. 

With an exception in a way, we postpone that decision to the person that sees that exception.

Reply:
However I am not strong on that. Just wanted to bring it up for a discussion.

Reply:
I see and I agree that this can be a point of concern. For now, I will leave it like this because the `translateForBatch/Streaming` is new, so hopefully whoever touches it, will know what he/she is doing. And as a last resort we use the `Transformation` setting as default. In the future we will see, as this may also change now that the `Translator` could become "thinner".